src/hotspot/os/windows/os_windows.cpp

```diff
@@ -3982,14 +3982,201 @@ bool   os::win32::_is_windows_server         = false;
+}
+
+bool os::win32::is_windows_server_2022_or_greater() {
+  initialize_windows_version();
+
+  // Windows Server 2022 starts at build 20348.169 as per
+  // https://learn.microsoft.com/en-us/windows/release-health/release-information
+  return (_major_version >= 10 && _build_number >= 20348 && IsWindowsServer());
+}
+
+int os::win32::bit_count(ULONG64 argument) {
+  int bitcount = 0;
+  while (argument != 0) {
+    argument &= (argument-1);
+    bitcount++;
+  }
+  return bitcount;
+}
+
+DWORD os::win32::active_processors_in_job_object() {
+  DWORD processors = 0;
+
+  LPVOID job_object_information = NULL;
+  DWORD job_object_information_length = 0;
+
+  if (!QueryInformationJobObject(NULL, JobObjectGroupInformationEx, NULL, 0, &job_object_information_length)) {
+    DWORD last_error = GetLastError();
+    if (last_error == ERROR_INSUFFICIENT_BUFFER) {
+      DWORD group_count = job_object_information_length / sizeof(GROUP_AFFINITY);
+
+      job_object_information = os::malloc(job_object_information_length, mtInternal);
+      if (job_object_information != NULL) {
+          if (QueryInformationJobObject(NULL, JobObjectGroupInformationEx, job_object_information, job_object_information_length, &job_object_information_length)) {
+            group_count = job_object_information_length / sizeof(GROUP_AFFINITY);
+
+            for (DWORD i = 0; i < group_count; i++) {
+              KAFFINITY group_affinity = ((GROUP_AFFINITY*)job_object_information)[i].Mask;
+              processors += bit_count(group_affinity);
+            }
+
+            assert(processors > 0, "Must find at least 1 logical processor");
+          } else {
+            warning("QueryInformationJobObject() failed: GetLastError->%ld.", GetLastError());
+          }
+
+          os::free(job_object_information);
+      } else {
+          warning("os::malloc() failed to allocate %ld bytes for QueryInformationJobObject", job_object_information);
+      }
+    }
+  }
+
+  return processors;
+}
+
+DWORD os::win32::system_logical_processor_count() {
+  DWORD logical_processors = 0;
+  typedef BOOL(WINAPI* LPFN_GET_LOGICAL_PROCESSOR_INFORMATION_EX)(
+      LOGICAL_PROCESSOR_RELATIONSHIP, PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX, PDWORD);
+
+  LPFN_GET_LOGICAL_PROCESSOR_INFORMATION_EX glpiex;
+
+  glpiex = (LPFN_GET_LOGICAL_PROCESSOR_INFORMATION_EX)GetProcAddress(
+      GetModuleHandle(TEXT("kernel32")),
+      "GetLogicalProcessorInformationEx");
+
+  if (glpiex != NULL) {
+    LOGICAL_PROCESSOR_RELATIONSHIP relationship_type = RelationGroup;
+    PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX sytem_logical_processor_info = NULL;
+    DWORD returned_length = 0;
+
+    // https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getlogicalprocessorinformationex
+    if (!glpiex(relationship_type, sytem_logical_processor_info, &returned_length)) {
+      DWORD last_error = GetLastError();
+
+      if (last_error == ERROR_INSUFFICIENT_BUFFER) {
```

Comment:
I'm seeing a pattern here! This must be the recommended way to do this.

Reply:
Yes, this is the way to use GetLogicalProcessorInformationEx. The above call can use nullptr for the 2nd argument to make it clearer though. I'll update that.